File: bwl.Imaging/FloatMatrix/CommonFloatMatrix.vb
Public Class CommonFloatMatrix
    Protected _matrices As List(Of Single(,))
    Protected _width As Integer
    Protected _height As Integer

    Public Sub New(channels As Integer, width As Integer, height As Integer)
        If channels < 1 Then Throw New ArgumentException("channels must be >0")
        If channels > 1024 Then Throw New ArgumentException("channels must be <=1024 due to memory overflow protection")
        If width < 1 Then Throw New ArgumentException("width must be >0")
        If height < 1 Then Throw New ArgumentException("height must be >0")
        _width = width
        _height = height
        _matrices = New List(Of Single(,))
        For i = 1 To channels
            Dim channel(width - 1, height - 1) As Single
            _matrices.Add(channel)
        Next
    End Sub

    Public ReadOnly Property Matrix(channel As Integer) As Single(,)
        Get
            Return _matrices(channel)
        End Get
    End Property

    Protected Function GetWidth(matrix As Single(,)) As Integer
        Return matrix.GetLength(0)
    End Function

    Protected Function GetHeight(matrix As Single(,)) As Integer
        Return matrix.GetLength(1)
    End Function

    Public Sub New(matrices As IEnumerable(Of Byte(,)), multiplier As Single)
        If matrices Is Nothing Then Throw New ArgumentException("matrices must not be null")
        If matrices.Count = 0 Then Throw New ArgumentException("matrices must contain at least one matrix")
        _width = matrices(0).GetLength(0)
        _height = matrices(0).GetLength(1)
        _matrices = New List(Of Single(,))

        For Each mtr In matrices
            If mtr.GetLength(0) <> _width Then Throw New Exception("all matrices must be one size")
            If mtr.GetLength(1) <> _height Then Throw New Exception("all matrices must be one size")
            Dim channel(Width - 1, Height - 1) As Single
            For x = 0 To _width - 1
                For y = 0 To _height - 1
                    channel(x, y) = CSng(mtr(x, y)) * multiplier
                Next
            Next
            _matrices.Add(channel)
        Next
    End Sub

    Public Sub New(matrices As IEnumerable(Of Single(,)))
        If matrices Is Nothing Then Throw New ArgumentException("matrices must not be null")
        If matrices.Count = 0 Then Throw New ArgumentException("matrices must contain at least one matrix")
        _width = GetWidth(matrices(0))
        _height = GetHeight(matrices(0))
        _matrices = New List(Of Single(,))

        For Each mtr In matrices
            If GetWidth(mtr) <> _width Then Throw New Exception("all matrices must be one size")
            If GetHeight(mtr) <> _height Then Throw New Exception("all matrices must be one size")
            _matrices.Add(mtr)
        Next
    End Sub

    Public ReadOnly Property Width As Integer
        Get
            Return _width
        End Get
    End Property

    Public ReadOnly Property Height As Integer
        Get
            Return _height
        End Get
    End Property

    Public Function CloneMatrix(matrix As Single(,)) As Single(,)
        Dim arr(Width - 1, Height - 1) As Single
        For x = 0 To Width - 1
            For y = 0 To Height - 1
                arr(x, y) = matrix(x, y)
            Next
        Next
        Return arr
    End Function

    Public Function ResizeMatrixHalf(matrix As Single(,)) As Single(,)
        Dim result(_width \ 2 - 1, _height \ 2 - 1) As Single

        For x = 0 To _width \ 2 - 1
            For y = 0 To _height \ 2 - 1
                Dim point As Integer = 0
                point += matrix(x * 2, y * 2)
                point += matrix(x * 2 + 1, y * 2)
                point += matrix(x * 2, y * 2 + 1)
                point += matrix(x * 2 + 1, y * 2 + 1)
                point \= 4
                result(x, y) = point
            Next
        Next
        Return result
    End Function

    Public Function ResizeMatrixTwo(matrix As Single(,)) As Single(,)
        Dim result(_width * 2 - 1, _height * 2 - 1) As Single
        For x = 0 To Width - 1
            For y = 0 To Height - 1
                result(x * 2, y * 2) = matrix(x, y)
                result(x * 2 + 1, y * 2) = matrix(x, y)
                result(x * 2, y * 2 + 1) = matrix(x, y)
                result(x * 2 + 1, y * 2 + 1) = matrix(x, y)
            Next
        Next
        Return result
    End Function

    Public Overridable Function Clone() As CommonFloatMatrix
        Dim list As New List(Of Single(,))
        For Each mtr In _matrices
            list.Add(CloneMatrix(mtr))
        Next
        Return New CommonFloatMatrix(list)
    End Function

    Public Overridable Function ResizeTwo() As CommonFloatMatrix
        Dim list As New List(Of Single(,))
        For Each mtr In _matrices
            list.Add(ResizeMatrixTwo(mtr))
        Next
        Return New CommonFloatMatrix(list)
    End Function

    Public Overridable Function ResizeHalf() As CommonFloatMatrix
        Dim list As New List(Of Single(,))
        For Each mtr In _matrices
            list.Add(ResizeMatrixHalf(mtr))
        Next
        Return New CommonFloatMatrix(list)
    End Function

End Class

File: Bwl.Imaging/Matrix/FloatMatrix/CommonFloatMatrix.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace Bwl.Imaging
{
    public class CommonFloatMatrix
    {
        protected List<double[]> _matrices;
        protected int _width;
        protected int _height;

        public CommonFloatMatrix(int channels, int width, int height)
        {
            if (channels < 1)
                throw new ArgumentException("channels must be >0");
            if (channels > 1024)
                throw new ArgumentException("channels must be <=1024 due to memory overflow protection");
            if (width < 1)
                throw new ArgumentException("width must be >0");
            if (height < 1)
                throw new ArgumentException("height must be >0");
            _width = width;
            _height = height;
            _matrices = new List<double[]>();
            for (int i = 1, loopTo = channels; i <= loopTo; i++)
            {
                var channel = new double[(width * height)];
                _matrices.Add(channel);
            }
        }

        public CommonFloatMatrix(IEnumerable<double[]> matrices, int width, int height)
        {
            if (matrices is null)
                throw new ArgumentException("matrices must not be null");
            if (matrices.Count() == 0)
                throw new ArgumentException("matrices must contain at least one matrix");
            _width = width;
            _height = height;
            _height = matrices.ElementAtOrDefault(0).Length / width;
            _matrices = new List<double[]>();

            foreach (var mtr in matrices)
            {
                if (mtr.Length != width * height)
                    throw new Exception("all matrices must have width*height elements");
                _matrices.Add(mtr);
            }
        }

        public CommonFloatMatrix(IEnumerable<int[]> matrices, int width, int height)
        {
            if (matrices is null)
                throw new ArgumentException("matrices must not be null");
            if (matrices.Count() == 0)
                throw new ArgumentException("matrices must contain at least one matrix");
            _width = width;
            _height = height;
            _matrices = new List<double[]>();

            foreach (var mtr in matrices)
            {
                if (mtr.Length != width * height)
                    throw new Exception("all matrices must have width*height elements");
                var channel = new double[mtr.Length];

                for (int i = 0, loopTo = mtr.Length - 1; i <= loopTo; i++)
                {
                    double pixel = mtr[i];
                    channel[i] = pixel;
                }
                _matrices.Add(channel);
            }
        }

        public CommonFloatMatrix(IEnumerable<double[]> matrices, int width, int height, double multiplier)
        {
            if (matrices is null)
                throw new ArgumentException("matrices must not be null");
            if (matrices.Count() == 0)
                throw new ArgumentException("matrices must contain at least one matrix");
            _width = width;
            _height = height;
            _matrices = new List<double[]>();

            foreach (var mtr in matrices)
            {
                if (mtr.Length != width * height)
                    throw new Exception("all matrices must have width*height elements");
                var channel = new double[mtr.Length];

                for (int i = 0, loopTo = mtr.Length - 1; i <= loopTo; i++)
                {
                    double pixel = mtr[i] * multiplier;
                    channel[i] = pixel;
                }
                _matrices.Add(channel);
            }
        }

        public CommonFloatMatrix(IEnumerable<int[]> matrices, int width, int height, double multiplier)
        {
            if (matrices is null)
                throw new ArgumentException("matrices must not be null");
            if (matrices.Count() == 0)
                throw new ArgumentException("matrices must contain at least one matrix");
            _width = width;
            _height = height;
            _matrices = new List<double[]>();

            foreach (var mtr in matrices)
            {
                if (mtr.Length != width * height)
                    throw new Exception("all matrices must have width*height elements");
                var channel = new double[mtr.Length];

                for (int i = 0, loopTo = mtr.Length - 1; i <= loopTo; i++)
                {
                    double pixel = mtr[i] * multiplier;
                    channel[i] = pixel;
                }
                _matrices.Add(channel);
            }
        }

        public double GetMatrixPixel(int channel, int x, int y)
        {
            return _matrices[channel][x + y * Width];
        }
        public void SetMatrixPixel(int channel, int x, int y, double value)
        {
            _matrices[channel][x + y * Width] = value;
        }

        public double[] GetMatrix(int channel)
        {
            return _matrices[channel];
        }

        public int FitX(int x)
        {
            if (x < 0)
                x = 0;
            if (x >= Width)
                x = Width - 1;
            return x;
        }

        public int FitY(int y)
        {
            if (y < 0)
                y = 0;
            if (y >= Height)
                y = Height - 1;
            return y;
        }

        public int HalfWidth()
        {
            return Width / 2;
        }

        public int HalfHeight()
        {
            return Height / 2;
        }

        public int Width
        {
            get
            {
                return _width;
            }
        }

        public int Height
        {
            get
            {
                return _height;
            }
        }

        public double[] CloneMatrix(double[] matrix)
        {
            var arr = new double[matrix.Length];
            Array.Copy(matrix, arr, matrix.Length);
            return arr;
        }

        public double[] ResizeMatrixHalf(double[] matrix)
        {
            var result = new double[(_width / 2 * (_height / 2))];
            for (int y = 0, loopTo = _height / 2 - 1; y <= loopTo; y++)
            {
                int lineOffset1 = y * 2 * _width;
                int lineOffset2 = (y * 2 + 1) * _width;
                int resOffset = y * (_width / 2);
                for (int x = 0, loopTo1 = _width / 2 - 1; x <= loopTo1; x++)
                {
                    double elem = matrix[x * 2 + lineOffset1];
                    elem += matrix[x * 2 + 1 + lineOffset1];
                    elem += matrix[x * 2 + lineOffset2];
                    elem += matrix[x * 2 + 1 + lineOffset2];
                    result[x + resOffset] = elem / 4d;
                }
            }
            return result;
        }

        public double[] ResizeMatrixTwo(double[] matrix)
        {
            var result = new double[(_width * 2 * _height * 2)];
            for (int y = 0, loopTo = Height - 1; y <= loopTo; y++)
            {
                int lineOffset1 = y * 2 * _width * 2;
                int lineOffset2 = (y * 2 + 1) * _width * 2;
                int offset = y * _width;
                for (int x = 0, loopTo1 = _width - 1; x <= loopTo1; x++)
                {
                    double elem = matrix[x + offset];
                    result[x * 2 + lineOffset1] = elem;
                    result[x * 2 + 1 + lineOffset1] = elem;
                    result[x * 2 + lineOffset2] = elem;
                    result[x * 2 + 1 + lineOffset2] = elem;
                }
            }
            return result;
        }

        public virtual CommonFloatMatrix Clone()
        {
            var list = new List<double[]>();
            foreach (var mtr in _matrices)
                list.Add(CloneMatrix(mtr));
            return new CommonFloatMatrix(list, Width, Height);
        }

        public virtual CommonFloatMatrix ResizeTwo()
        {
            var list = new List<double[]>();
            foreach (var mtr in _matrices)
                list.Add(ResizeMatrixTwo(mtr));
            return new CommonFloatMatrix(list, Width * 2, Height * 2);
        }

        public virtual CommonFloatMatrix ResizeHalf()
        {
            var list = new List<double[]>();
            foreach (var mtr in _matrices)
                list.Add(ResizeMatrixHalf(mtr));
            return new CommonFloatMatrix(list, Width / 2, Height / 2);
        }
    }
}